replace usages of snprintf, sscanf

Mrmaxmeier opened the pull request:
This closes #192

Mrmaxmeier committed:
replace usages of snprintf, sscanf

cormacrelf commented on dpx/src/dpx_pdfobj.rs:
```diff
@@ -3753,13 +3750,16 @@ pub unsafe extern "C" fn pdf_open(
+                        .unwrap_or(0)
+                }
+            }
+            if minor == 0 {
```
Isn’t 0 a valid value for minor version? The original code checked parse failure, so that would require checking Err from parse().

cormacrelf replied:
I mean 1.0 is not supported (see above) but this code will return different results than the original and report “broken” when it should report “unsupported”.

crlf0710 replied:
yup, 1.0 is valid value(though useless in practice) , and 2.0 is valid too by the way:)

cormacrelf commented on dpx/src/specials/misc.rs:
```diff
@@ -19,30 +19,44 @@
+        parts.push(&elem[1..]);
+    }
+
+    if parts.len() != 3 {
```
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=80669d181c564d1349bc380b8e2a2beb

There are four parts, the last one is empty. So it will fail on the starts_with.

cormacrelf reviewed (commented):
GitHub is forcing me to leave a comment

Mrmaxmeier committed:
silence/fix build warnings for engine/*

Mrmaxmeier committed:
port tpic points to Vec

Mrmaxmeier committed:
port page_ranges to Vec

Mrmaxmeier committed:
refactor tt_get_positive_quad

fastmod -m -d . --extensions rs \
  "(tt_get_positive_quad\(\s+.+?,\s+)b(\".*?)\\\\x00.*?,(\s+)b(\".*?)\\\\x00.*?," \
  "\${1}\${2}\",\${3}\${4}\","

Mrmaxmeier committed:
remove unused dpx_create_temp_file

Mrmaxmeier committed:
correct "libpng-sys bug" comment

Mrmaxmeier committed:
0 as *const _ => ptr::null()

Mrmaxmeier committed:
0 as *mut _ => ptr::null_mut()

Mrmaxmeier committed:
remove unused imports

Mrmaxmeier committed:
dpx: drop extern "C"

Mrmaxmeier committed:
dpx: drop #[no_mangle]

Mrmaxmeier committed:
bibtex: drop `extern "C"`

Mrmaxmeier committed:
rustify some sprintfs

Mrmaxmeier committed:
fix scanf parsing regressions

Mrmaxmeier committed:
dpx: remove xrealloc

Mrmaxmeier committed:
pdf_add_stream with const to pdf_add_stream_str

fastmod -m -d . --extensions rs \
  "pdf_add_stream\(\s*([^\n]*?),\s+?b(\".*?)\\\\x00\".*?,\s+.*?\)"
  "pdf_add_stream_str(\${1}, \${2}\")"

Mrmaxmeier committed:
port loaded_fonts and def_fonts to Vec

Mrmaxmeier committed:
parse_pdf_version: split at all whitespace

This allows for exotic PDF with this header:

00000000: 2550 4446 2d31 2e34 204d 6172 7665 6c6c  %PDF-1.4 Marvell
00000010: 2053 656d 6963 6f6e 6475 6374 6f72 0a31   Semiconductor.1

cormacrelf commented:
Maybe it’s a good idea to add some tests for the version parser? It looks better but it’s hard to tell from just looking. Would help if you wanted to change it to nom later as well.

Mrmaxmeier committed:
fix macos build

burrbull commented:
I like these changes, but not sure about `ptr::null()`. It hides variable type and can complicate future refactoring.

Same also with `PdfObjRef` in #176 .

Mrmaxmeier committed:
trie_node: allow wrapping add

With checked (i32) adds, this panicked during format generation.

Mrmaxmeier commented:
This now includes the commits from #191 that got lost in #192.
I'll drop a6e7530 and 697712f if we decide to prefer `0 as *const _` over `ptr::null()`

burrbull commented:
If it is already in master, I'm not against.

crlf0710 commented:
Personally i think `ptr::null()` is fine at this stage ... and they'll disappear when we convert them into references.

cormacrelf commented:
Doesn’t every c2rust-generated “0 as *const _” have a corresponding explicit type annotation anyway, even in method bodies? I have been keeping the annotation but replacing the 0 with ptr::null.

crlf0710 commented:
Seems good. Just need to fix compile errors on Mac (seems extern types doesn't work well together with ptr::null()?) and Windows (are we still calling vsprintf from somewhere?), and we're good to go~

Mrmaxmeier committed:
remove unused dpx_warning

Mrmaxmeier committed:
re-introduce `xetex_snprintf` to fix windows linkage
